Add Interview model and routes

AndresNeisa opened the pull request:
I would appreciate a review on the Interview routes, I've been trying to use RESTful routes.

AndresNeisa committed:
Add interview model and routes

AndresNeisa committed:
Merge branch 'dev' into interview-model-routes

AndresNeisa committed:
Merge remote-tracking branch 'origin/BE-login-signup' into interview-model-routes

AndresNeisa committed:
Remove unused imports

AndresNeisa committed:
Merge branch 'dev' into interview-model-routes

AndresNeisa committed:
Protect the interview routes

The authorization middleware has been added to the interview routes.
Participants field has been replaced by owner and guest fields in the Interview model.

dproc96 commented on server/controllers/interview.js:
```diff
@@ -0,0 +1,109 @@
+// connect to the Interview Model
+let Interview = require('../models/interview');
+let User = require('../models/user');
```
These can be const

AndresNeisa replied:
The variables that don't require "let" have been changed to "const"

dproc96 commented on server/controllers/interview.js:
```diff
@@ -0,0 +1,109 @@
+	const user = await User.findOne({ email: req.user.user }, 'id');
+
+	let newInterview = Interview({
+		owner: user,
+		guest: guest,
```
These should both be ObjectIds correct? At the moment it looks like user is a Mongo document and guest is a string

AndresNeisa replied:
Yes, those should be ObjectIds. Right before assigning those values, I am getting the user id using the following command:
```
const user = await User.findOne({ email: req.user.user }, 'id');
```
I just get rid of the guest as it is not possible from the user's perspective to create an interview with a guest. To be able to add a guest to the interview the (.../api/guest/:id) endpoint should be reached

dproc96 commented on server/models/interview.js:
```diff
@@ -0,0 +1,37 @@
+			type: Boolean,
+			default: false,
+		},
+		difficult: {
```
```suggestion
		difficulty: {
```
I would change this name, because my intuition told me a variable called `difficult` would be a boolean.

AndresNeisa replied:
The field name has been updated as requested.

dproc96 commented on server/models/interview.js:
```diff
@@ -0,0 +1,37 @@
+
+let Interview = new mongoose.Schema(
+	{
+		owner: {
+			type: mongoose.ObjectId,
+			required: true,
+		},
```
For a reference to another model, you have to [name the model in the ref property](https://mongoosejs.com/docs/populate.html)

AndresNeisa replied:
References to owner and guest id have been added.

dproc96 commented on server/controllers/interview.js:
```diff
@@ -0,0 +1,109 @@
+		difficult: difficult,
+	});
+
+	Interview.create(newInterview, (err) => {
```
You're mixing create methods here, I think this would technically work but Interview.create expects an object or array rather than an Interview instance. You can use newInterview.save() for an instance.

AndresNeisa replied:
The method has been changed as requested

dproc96 commented on server/controllers/interview.js:
```diff
@@ -0,0 +1,109 @@
+	const { guest, date, theme, questions, difficult } = req.body;
+	const user = await User.findOne({ email: req.user.user }, 'id');
+
+	let newInterview = Interview({
```
```suggestion
	const newInterview = new Interview({
```

dproc96 commented on server/controllers/interview.js:
```diff
@@ -0,0 +1,109 @@
+	});
+};
+
+module.exports.updateInterview = (req, res, next) => {
```
(optional) It would be worth adding a check to all of these controllers that the user accessing the route is the owner of the specified interview (or guest if they have permissions too)

AndresNeisa replied:
We talk about that in the sprint planning meeting and we decided that we will keep those routes as they are for now

dproc96 reviewed (commented):
Had a couple of comments, also it looks like your Prettier is set to 4 space indent instead of 2 space indent, is that aligned with your group?

AndresNeisa commented:
I just made a pull request to solve the formatting issues. In there you can find a prettier config file, I hope that it solves all the formatting issues. 
Here is a link to the PR: 
https://github.com/hatchways/team-hulk/pull/42

AndresNeisa committed:
Add review changes

AndresNeisa committed:
Merge branch 'dev' into interview-model-routes

dproc96 commented on server/controllers/interview.js:
```diff
@@ -0,0 +1,111 @@
+const User = require("../models/user");
+
+module.exports.createInterview = async (req, res, next) => {
+  const { guest, date, theme, questions, difficulty } = req.body;
```
If you're removing guest from the creation then you should also remove it here

AndresNeisa replied:
Guest has been removed as requested.

dproc96 commented on server/controllers/interview.js:
```diff
@@ -0,0 +1,111 @@
+
+module.exports.createInterview = async (req, res, next) => {
+  const { guest, date, theme, questions, difficulty } = req.body;
+  const user = await User.findOne({ email: req.user.user }, "id");
+
+  const newInterview = new Interview({
+    owner: user,
```
Here you should have the user._id coming from the auth middleware correct? If not that should be added there rather than calling findOne here

AndresNeisa replied:
The middleware is now sending the userId, it was achieved by modifying signin.js. I would appreciate a review of that file as I am not sure if I used best practices there.

dproc96 commented on server/models/interview.js:
```diff
@@ -0,0 +1,39 @@
+      type: String,
+    },
+    questions: {
+      type: [mongoose.ObjectId],
```
You should add a ref to this property as well

AndresNeisa replied:
The Questions model has not been added to the database, adding a reference pointing to an unexisting collection won't generate errors?

dproc96 commented on server/routes/interview.js:
```diff
@@ -0,0 +1,27 @@
+let express = require('express');
+let router = express.Router();
+
+let auth = require('../middleware/authenticateJWT');
+
+// Connect to Interview controller
+let interviewController = require('../controllers/interview');
```
Whenever possible, it's preferred to use const instead of let

AndresNeisa replied:
I will have that in mind in future commits, the change has already been made.

dproc96 reviewed (commented):
Had a few more comments, but nice work addressing the previous review

AndresNeisa committed:
Add review comments

The auth token now carries the user id allowing the middleware to send back the id without having to query the db in the interview controller

AndresNeisa committed:
Merge branch 'dev' into interview-model-routes

AndresNeisa committed:
Add interview documentation
